### ISSUE TITLE
Express Check-In Routes

### ISSUE THREAD

**Issue by steets250** [acmucsd/membership-portal#107]
I'm currently working on an express check-in feature and would like a couple of new routes to be added to the backend.

**Feature Overview**: The user will scan a QR code taking them to something like: acmucsd.com/checkin?code=t3sting. If signed in, they will receive points for the event and see a confirmation page. If not signed in, they'll have the option to sign in, register, or express register. The express registration requires only their email, so we'll need a way of creating a partially complete user profile that can be finished at a later time.

Here's the Figma link if you want to see a bit more clearly: https://www.figma.com/file/hTnqyyJN8QUQsLaz4Q3rzk/Dev-Jam-Doc

**New Routes**:
- /event GET: Given an attendanceCode, return public event info if match found. (Similar to /attendance but non-authenticated)
- /user POST: Given an event and an email, create a new user and award them points for the event. The user will need some sort of special "partial" state (different from PENDING), as we'll check for this and ensure that signup is completed later on before they can fully access the portal.
- /user PATCH: Similar to the existing /user patch, but will need to require all fields and set the user's state to ACTIVE.

**Comment by sumeet-bansal** [comment]
I probably missed the discussion, but is this form of express check-in really necessary? The only fields the user would be skipping are name, password, major, and year, and the current registration process seems pretty "express" already. The portal's supposed to make it easier to see the breakdowns for who's attending what kind of events, and allowing users to check-in with no listed major or year seems counterproductive, especially since it takes less than half a minute to fill those extra fields out. Express check-in using NFC tags or something is a different discussion since that simplifies the check-in process for members, but this doesn't seem to speed up check-in much and makes the portal less effective by weakening our guarantees of knowing the basics about our attendees.

**Comment by sumeet-bansal** [comment]
That being said, a QR code that takes you to the registration/login page and automatically checks you in once you're authenticated would definitely be great, but that could be done entirely in the FE, and I think this version of express check-in could really diminish the effectiveness of the portal unless there's some way to force a user to supply major/year after check-in (but at that point, you might as well just go through the regular registration process).

**Comment by steets250** [comment]
@sumeet-bansal I agree that the express registration part isn't really critical, but it was kept in since the goal was to make check-ins as fast as possible for new members and minimize time spent on the phone. I'll go ahead and just use the existing login and register processes. Can you still add the ability to retrieve an event given a check in code, when not signed in? That's the only thing that I think will be needed to add this functionality to the front end.

**Comment by sumeet-bansal** [comment]
I still disagree that check-ins need to be faster, but if that's something that was already discussed, I'm not gonna argue further.

Re: an unauthenticated endpoint for retrieving an event given a check-in code, how soon do you need it? We're currently working on the portal rewrite (it's almost done minus the merch store, just needs "everything else" like a CircleCI config) but this should be a pretty quick PR if you want it to take a stab at it yourself. It could just add a check-in code filter to the `GET /events` search endpoint I think, but to make sure people don't try brute-forcing attendance codes, it'll need to be rate-limited (although I'd still prefer it to be authenticated if possible to really discourage any sort of abuse). The other email endpoints need to be rate-limited as well though, so we can skip that for now and implement it after the rewrite.

CC @shravanhariharan2.

**Comment by steets250** [comment]
It's not urgent at all. I'll give the event code route a try since it's pretty similar to the existing endpoint.
